static/js/components/Editor.js

```diff
@@ -0,0 +1,353 @@
+// @flow
+import React from "react"
+import R from "ramda"
+import { connect } from "react-redux"
+import { EditorState } from "prosemirror-state"
+import { EditorView } from "prosemirror-view"
+import { exampleSetup } from "prosemirror-example-setup"
+import {
+  schema,
+  defaultMarkdownParser,
+  defaultMarkdownSerializer
+} from "prosemirror-markdown"
+import { toggleMark, wrapIn } from "prosemirror-commands"
+import { wrapInList } from "prosemirror-schema-list"
+
+import AddLinkMenu from "../components/AddLinkMenu"
+
+import { makeUUID } from "../lib/util"
+import { configureForm } from "../lib/forms"
+import { markIsActive } from "../lib/prosemirror"
+
+import type { Dispatch } from "redux"
+
+// helpers for the menu bar
+type MenuBarItem = {
+  command: Function,
+  title: string,
+  icon: React$Element<"i">,
+  active?: Function
+}
+
+// each array in this array represents a 'group' of buttons
+// (we want to group certain buttons together with a border
+// around them in the UI)
+export const menuBarManifest: Array<Array<MenuBarItem>> = [
+  [
+    {
+      command: toggleMark(schema.marks.strong),
+      title:   "Bold",
+      icon:    <i className="material-icons format_bold">format_bold</i>,
+      active:  markIsActive(schema.marks.strong)
+    }
+  ],
+  [
+    {
+      command: toggleMark(schema.marks.em),
+      title:   "Italic",
+      icon:    <i className="material-icons format_italic">format_italic</i>,
+      active:  markIsActive(schema.marks.em)
+    }
+  ],
+  [
+    {
+      command: wrapInList(schema.nodes.bullet_list),
+      title:   "bullet list",
+      icon:    (
+        <i className="material-icons format_list_bulleted">
+          format_list_bulleted
+        </i>
+      )
+    },
+    {
+      command: wrapInList(schema.nodes.ordered_list),
+      title:   "ordered list",
+      icon:    (
+        <i className="material-icons format_list_numbered">
+          format_list_numbered
+        </i>
+      )
+    },
+    {
+      command: wrapIn(schema.nodes.blockquote),
+      title:   "quote",
+      icon:    <i className="material-icons format_quote">format_quote</i>
+    }
+  ]
+]
+
+export const newLinkForm = () => ({
+  url:  "",
+  text: ""
+})
+
+type Props = {
+  onChange: (serialized: string) => void,
+  initialValue?: string,
+  autoFocus?: boolean,
+  dispatch: Dispatch<*>
+}
+
+type State = {
+  editorState: Object,
+  showLinkMenu: boolean
+}
+
+export class Editor extends React.Component<Props, State> {
+  node: { current: null | React$ElementRef<typeof HTMLDivElement> }
+  view: Object
+  uuid: string
+  formHelpers: Object
+  dispatchTransaction: Function
```

Comment:
yeah it's a bit weird, if you look you'll see that `dispatchTransaction` is a class method (rather than a class field, which we normally do). to get `this` to be bound properly I then have to bind `dispatchTransaction` in the constructor, and Flow was complaining about that property not having a type annotation on the class, so I had to add this annotation to get it to be quiet.

I had to do all that silliness in the first place so that I could intercept the `dispatchTransaction` method with a `sinon` stub. I wanted to do this to check that it was being properly passed into the `EditorView` object in a test. The issue, though, was that class field type methods (which we normally use, declaring them with an arrow function) are not defined on the class' prototype in the same way that a class method is, so there is not a way to stub them out before instantiating the react component.

With enzyme it's possible to stub out a class-field type method after instantiating the component, using `wrapper.instance().methodName` to grab a reference to the method, but I couldn't do that in this particular case because we pass `this.dispatchTransaction` to the `EditorView` instance which lives at `this.view` in the `componentDidMount` callback. This meant that, if I mounted the component, the `EditorView` object was already holding on to a reference to the un-stubbed `dispatchTransaction` method, and stubbing out the method on `.instance()` didn't have any effect. So kind of a chicken-and-egg situation, and the solution was to move to a class method, so it can be stubbed out on `Editor.prototype` before instantiating the class, so that when the `EditorView` is created it's holding on to a reference to the stubbed method, instead of the real one.